File: src/main/cc/wfa/virtual_people/common/field_filter/utils/values_parser_test.cc
// Copyright 2021 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "wfa/virtual_people/common/field_filter/utils/values_parser.h"

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "gmock/gmock.h"
#include "google/protobuf/descriptor.h"
#include "google/protobuf/message.h"
#include "gtest/gtest.h"
#include "src/main/cc/wfa/virtual_people/common/field_filter/test/test.pb.h"
#include "src/test/cc/testutil/matchers.h"
#include "src/test/cc/testutil/status_macros.h"

namespace wfa_virtual_people {
namespace {

using ::testing::UnorderedElementsAre;
using ::wfa::StatusIs;
using ::wfa_virtual_people::test::TestProtoB;

TEST(ValuesParserTest, TestInt32) {
  ASSERT_OK_AND_ASSIGN(
      ParsedValues<int32_t> parser, ParseValues<int32_t>("1,2,1"));
  EXPECT_THAT(parser.values, UnorderedElementsAre(1, 2));
}

TEST(ValuesParserTest, TestInt32Invalid) {
  EXPECT_THAT(
      ParseValues<int32_t>("1,a,1").status(),
      StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(ValuesParserTest, TestInt64) {
  ASSERT_OK_AND_ASSIGN(
      ParsedValues<int64_t> parser, ParseValues<int64_t>("1,2,1"));
  EXPECT_THAT(parser.values, UnorderedElementsAre(1, 2));
}

TEST(ValuesParserTest, TestInt64Invalid) {
  EXPECT_THAT(
      ParseValues<int64_t>("1,a,1").status(),
      StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(ValuesParserTest, TestUInt32) {
  ASSERT_OK_AND_ASSIGN(
      ParsedValues<uint32_t> parser, ParseValues<uint32_t>("1,2,1"));
  EXPECT_THAT(parser.values, UnorderedElementsAre(1, 2));
}

TEST(ValuesParserTest, TestUInt32Invalid) {
  EXPECT_THAT(
      ParseValues<uint32_t>("1,a,1").status(),
      StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(ValuesParserTest, TestUInt64) {
  ASSERT_OK_AND_ASSIGN(
      ParsedValues<uint64_t> parser, ParseValues<uint64_t>("1,2,1"));
  EXPECT_THAT(parser.values, UnorderedElementsAre(1, 2));
}

TEST(ValuesParserTest, TestUInt64Invalid) {
  EXPECT_THAT(
      ParseValues<uint64_t>("1,a,1").status(),
      StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(ValuesParserTest, TestBool) {
  ASSERT_OK_AND_ASSIGN(
      ParsedValues<bool> parser, ParseValues<bool>("true,false,true"));
  EXPECT_THAT(parser.values, UnorderedElementsAre(true, false));
}

TEST(ValuesParserTest, TestBoolInvalid) {
  EXPECT_THAT(
      ParseValues<bool>("true,a,false").status(),
      StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(ValuesParserTest, TestString) {
  ASSERT_OK_AND_ASSIGN(
      ParsedValues<const std::string&> parser,
      ParseValues<const std::string&>("a,b,a"));
  EXPECT_THAT(parser.values, UnorderedElementsAre("a", "b"));
}

TEST(ValuesParserTest, TestEnum) {
  ASSERT_OK_AND_ASSIGN(
      ParsedValues<const google::protobuf::EnumValueDescriptor*> parser_1,
      ParseEnumValues(TestProtoB::TestEnum_descriptor(), "1,2,1"));
  EXPECT_THAT(parser_1.values, UnorderedElementsAre(1, 2));

  ASSERT_OK_AND_ASSIGN(
      ParsedValues<const google::protobuf::EnumValueDescriptor*> parser_2,
      ParseEnumValues(TestProtoB::TestEnum_descriptor(),
                      "TEST_ENUM_1,TEST_ENUM_2,TEST_ENUM_1"));
  EXPECT_THAT(parser_2.values, UnorderedElementsAre(1, 2));
}

TEST(ValuesParserTest, TestEnumInvalid) {
  EXPECT_THAT(
      ParseEnumValues(TestProtoB::TestEnum_descriptor(), "1,a,1").status(),
      StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

}  // namespace
}  // namespace wfa_virtual_people

File: src/test/kotlin/org/wfanet/virtualpeople/common/fieldfilter/utils/ValuesParserTest.kt
// Copyright 2022 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.wfanet.virtualpeople.common.fieldfilter.utils

import java.lang.NumberFormatException
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.wfanet.virtualpeople.common.test.TestProtoB

@RunWith(JUnit4::class)
class ValuesParserTest {

  @Test
  fun `value parser valid cases`() {
    assertEquals(parseValue<Int>("1,2,1"), setOf(1, 2))
    assertEquals(parseValue<Long>("1,2,1"), setOf(1L, 2L))
    assertEquals(parseValue<Boolean>("true,true,false"), setOf(true, false))
    assertEquals(parseValue<String>("a,b,a"), setOf("a", "b"))
    assertEquals(parseValue<String>(""), setOf(""))
    assertEquals(parseValue<String>(",a"), setOf("", "a"))
    assertEquals(parseValue<String>("\"a,b,a"), setOf("\"a", "a", "b"))
  }

  @Test
  fun `value parser invalid cases`() {
    assertFailsWith<NumberFormatException> { parseValue<Int>("1,a,1") }
    assertFailsWith<NumberFormatException> { parseValue<Long>("1,a,1") }
    val exception = assertFailsWith<IllegalArgumentException> { parseValue<Boolean>("true,a") }
    assertTrue(exception.message!!.contains("The string doesn't represent a boolean value"))
  }

  @Test
  fun `enum value parser valid cases`() {
    val descriptor = TestProtoB.TestEnum.getDescriptor()
    assertEquals(parseEnumValues(descriptor, "1,2,1"), setOf(1, 2))
    assertEquals(parseEnumValues(descriptor, "TEST_ENUM_1,TEST_ENUM_2,TEST_ENUM_1"), setOf(1, 2))
    assertEquals(parseEnumValues(descriptor, "1,TEST_ENUM_2,TEST_ENUM_1,2"), setOf(1, 2))
  }

  @Test
  fun `enum value parser invalid cases`() {
    val descriptor = TestProtoB.TestEnum.getDescriptor()
    val exception = assertFailsWith<IllegalStateException> { parseEnumValues(descriptor, "1,a,1") }
    assertTrue(exception.message!!.contains("Not a valid enum name or integer"))
  }
}
